python/pysrc/kaskada/sources/arrow.py

```diff
@@ -109,15 +105,61 @@ def __init__(
         )
 
         self._convert_options = pyarrow.csv.ConvertOptions(column_types=schema)
-        self.add_rows(rows)
 
-    def add_rows(self, rows: dict | list[dict]) -> None:
+    @staticmethod
+    async def create(
+        *,
+        time_column: str,
+        key_column: str,
+        retained: bool = True,
+        rows: Optional[dict | list[dict]] = None,
+        subsort_column: Optional[str] = None,
+        schema: Optional[pa.Schema] = None,
+        grouping_name: Optional[str] = None,
+        time_unit: Optional[TimeUnit] = None) -> PyDict:
+        """Create a source reading from rows represented as dicts.
+
+        Args:
+            time_column: The name of the column containing the time.
+            key_column: The name of the column containing the key.
+            retained: Whether added rows should be retained for queries.
+              If True, rows (both provided to the constructor and added later) will be retained
+              for interactive queries. If False, rows will be discarded after being sent to any
+              running materializations. Consider setting this to False when the source will only
+              be used for materialization to avoid unnecessary memory usage.
+            subsort_column: The name of the column containing the subsort.
+              If not provided, the subsort will be assigned by the system.
+            schema: The schema to use. If not provided, it will be inferred from the input.
+            grouping_name: The name of the group associated with each key.
+              This is used to ensure implicit joins are only performed between data grouped
+              by the same entity.
+            time_unit: The unit of the time column. One of `ns`, `us`, `ms`, or `s`.
+              If not specified (and not specified in the data), nanosecond will be assumed.
+        """
+        if schema is None:
+            if rows is None:
+                raise ValueError("Must provide schema or rows")
+            schema = pa.Table.from_pylist(rows).schema
```

Comment:
I think this was mentioned in a conversation earlier but -- would it be possible to defer schema inference to when rows _are_ added?

Reply:
Consider:

```
source = kd.sources.PyDict(...)
source.col("x") # does this throw an error? we don't know
source.add_row({ y: 5 }) # does this throw an error?
```

Reply:
> `source.col("x") # does this throw an error? we don't know`

Yes: "Source has not defined a schema" 

> `source.add_row({ y: 5 }) # does this throw an error?`

No, but this sets the source schema to `{y: int}`

Just an idea -- I'm sure there may be examples that make it confusing, and I think going explicit for now isn't a bad idea.

Reply:
> Yes: "Source has not defined a schema"

So this forces you to add a row before you use it:

```
source = kd.sources.PyDict(...)
source.add_row({ y: 5 })
source.col("x")
```

This is just longer and more error-prone than allowing it to be combined like below:

```
source = kd.sources.PyDict({ y: 5 }, ...)
source.col("x")
```